### ISSUE TITLE
Detached elements & memory leak?

### ISSUE THREAD

**Issue by code-div** [hotwired/turbo#855]
I've been working on implementing Turbo on a site that unfortunately requires the use of quite a few 3rd party libraries/widgets (Gtag, UET, Trustpilot, Facebook etc). Most of which appear to cause major memory leaks as they contain code that retains references to the entire body element every time the user visits a new page. I'm talking of the JS heap size growing many-fold in just a few visits around the site. I can see copies of body pile up when inspecting using the Edge dev tools' Detached Elements pane.

Any such leaks in my own code I can handle with ease by releasing references & event handlers at the right time, but plugging the leaks from 3rd parties is proving extremely difficult and unreliable. 

Am I really supposed to be "hacking" into each of these libs to destroy these references? Or could there be a genuine issue that may need addressing on Turbo side? I haven't seen anyone else talk about this, suggesting I may be doing something wrong instead.

**Comment by tleish** [comment]
Can you provide a more concrete example of the code used to load and instantiate one of the libraries?

**Comment by code-div** [comment]
The libs are instantiated in the standard ways according to the respective dev's instructions. Most common example would be Google Analytics included via Google Tag Manager. I finally found others experiencing the same problem in a SPA setting e.g. https://www.voorhoede.nl/en/blog/your-website-probably-has-a-memory-leak/ and used similar solutions for all the libs I was dealing with.

Couple other, more tricky memory leaks were from 3rd party tools using iframes that cannot be reloaded on Turbo visits. These got solved once I got Turbo to behave more like a "conventional SPA", using the patch from https://github.com/hotwired/turbo/issues/499. This meant the iframes could stay in <body> forever, while Turbo would load the new body into a separate container. Later I heard similar functionality could be present in a future version of Turbo (which would be very welcome) : https://github.com/hotwired/turbo/pull/711

So in the end I realised these memory leaks are not really a Turbo problem, but a simple harsh reality when working with SPA and 3rd party widgets/tools not necessarily designed with SPAs in mind.
